# 01 - Load and Understand Data

## Objective

Loaded all the seven raw datasets and tried to understand their structure, grain,
columns, data types, timestamps and basic characteristics before
performing any cleaning or analysis.

In [2]:
import pandas as pd

In [7]:
captains = pd.read_csv("../data/captains.csv")
captains

,captain_id,signup_ts,city,vehicle_type,acquisition_channel,signup_zone_id,device_tier,app_language,age_band
0,CPT104359,2026-01-01 06:08:00,Pune,Auto,organic_app,PUN-Z23,low,kn,25-34
1,CPT120008,2026-01-01 06:52:00,Pune,Cab,referral,PUN-Z05,mid,mr,45+
2,CPT102049,2026-01-01 07:02:00,Hyderabad,Cab,organic_app,HYD-Z01,high,te,35-44
3,CPT123006,2026-01-01 07:15:00,Pune,Cab,organic_app,PUN-Z11,high,hi,25-34
4,CPT102642,2026-01-01 07:27:00,Bangalore,ERickshaw,gc_telecalling,BAN-Z06,low,hi,18-24
...,...,...,...,...,...,...,...,...,...
24995,CPT110215,2026-06-30 21:42:00,Delhi,Cab,referral,DEL-Z24,low,en,25-34
24996,CPT103759,2026-06-30 21:47:00,Pune,Cab,paid_digital,PUN-Z23,mid,kn,35-44
24997,CPT119842,2026-06-30 21:51:00,Hyderabad,Auto,organic_app,HYD-Z06,high,en,18-24
24998,CPT114818,2026-06-30 21:52:00,Bangalore,Cab,organic_app,BAN-Z17,mid,mr,25-34


In [9]:
doc_events = pd.read_csv("../data/doc_events.csv")
doc_events

,event_id,captain_id,doc_type,attempt_no,event_type,event_ts,failure_reason
0,EV000000000,CPT104359,DL,1,upload_success,2026-01-01 20:30:06.229672479,NaN
1,EV000000001,CPT104359,DL,1,verification_pass,2026-01-01 23:26:54.472703731,NaN
2,EV000000002,CPT104359,RC,1,upload_success,2026-01-02 09:36:49.336302871,NaN
3,EV000000003,CPT104359,RC,1,verification_pass,2026-01-02 12:41:51.324092998,NaN
4,EV000000004,CPT104359,AADHAAR,1,upload_success,2026-01-02 14:06:20.401070690,NaN
...,...,...,...,...,...,...,...
186277,EV000189009,CPT118784,DL,1,verification_pass,2026-06-30 23:21:57.734051686,NaN
186278,EV000189024,CPT112888,DL,1,upload_success,2026-06-30 22:30:32.437018572,NaN
186279,EV000189032,CPT108228,DL,1,upload_success,2026-06-30 21:46:56.507889270,NaN
186280,EV000189160,CPT113807,DL,1,upload_success,2026-06-30 23:41:12.099159103,NaN


In [10]:
approvals = pd.read_csv("../data/approvals.csv")
approvals

,captain_id,decision_ts,final_status,last_stage_reached,docs_cleared
0,CPT104359,2026-01-06 08:11:48.775157583,approved,NaN,6
1,CPT120008,NaN,dropped_in_docs,AADHAAR,2
2,CPT102049,NaN,dropped_in_docs,PERMIT,3
3,CPT123006,NaN,dropped_in_docs,INSURANCE,5
4,CPT102642,NaN,dropped_in_docs,PERMIT,3
...,...,...,...,...,...
24995,CPT110215,NaN,dropped_in_docs,INSURANCE,5
24996,CPT103759,NaN,dropped_in_docs,AADHAAR,2
24997,CPT119842,NaN,dropped_in_docs,AADHAAR,2
24998,CPT114818,NaN,dropped_in_docs,INSURANCE,5


In [11]:
activation = pd.read_csv("../data/activation.csv")
activation

,captain_id,first_order_ts,orders_d7,orders_d30,online_hours_d30
0,CPT104359,2026-01-11 12:11:54.524199600,7.0,33.0,26.9
1,CPT105916,2026-01-08 04:16:11.743250188,15.0,48.0,29.1
2,CPT115782,2026-01-10 11:14:57.939110321,9.0,31.0,20.2
3,CPT124093,2026-01-09 16:35:45.322422039,8.0,13.0,11.4
4,CPT101207,2026-01-09 18:43:22.114690071,4.0,9.0,6.1
...,...,...,...,...,...
4201,CPT100649,NaN,NaN,NaN,NaN
4202,CPT105704,2026-06-29 13:11:18.229384319,NaN,NaN,NaN
4203,CPT101859,NaN,NaN,NaN,NaN
4204,CPT112576,2026-06-30 10:18:21.460066786,NaN,NaN,NaN


In [12]:
nudges = pd.read_csv("../data/nudges.csv")
nudges

,captain_id,campaign_id,channel,sent_ts,delivered,clicked
0,CPT104359,CAMP_SMS_004,sms,2026-01-02 01:31:34.373244464,1,0
1,CPT102049,CAMP_WA_002,whatsapp,2026-01-02 20:02:31.634258914,1,0
2,CPT123006,CAMP_WA_002,whatsapp,2026-01-03 11:32:05.154545929,1,1
3,CPT105916,CAMP_WA_002,whatsapp,2026-01-05 01:25:34.559467887,1,1
4,CPT107181,CAMP_WA_001,whatsapp,2026-01-01 23:10:49.644254087,1,0
...,...,...,...,...,...,...
16309,CPT118083,CAMP_CALL_009,outbound_call,2026-06-30 19:41:17.131127305,1,1
16310,CPT119496,CAMP_WA_001,whatsapp,2026-06-30 22:59:51.375532451,1,0
16311,CPT117388,CAMP_SMS_004,sms,2026-06-30 16:27:52.459881139,1,0
16312,CPT110654,CAMP_CALL_009,outbound_call,2026-06-30 23:05:28.326161124,1,1


In [13]:
airport_hourly = pd.read_csv("../data/airport_hourly.csv")
airport_hourly

,zone_id,zone_type,hour_ts,requests,fulfilled_requests,unfulfilled_requests,online_captains,avg_eta_min,avg_surge_multiplier
0,APT-T1,airport_terminal,2026-05-01 00:00:00,83,31,52,15,9.13,2.04
1,APT-T1,airport_terminal,2026-05-01 01:00:00,111,27,84,10,10.13,2.17
2,APT-T1,airport_terminal,2026-05-01 02:00:00,86,21,65,13,10.89,2.19
3,APT-T1,airport_terminal,2026-05-01 03:00:00,54,25,29,15,9.27,1.80
4,APT-T1,airport_terminal,2026-05-01 04:00:00,36,34,2,17,4.54,1.08
...,...,...,...,...,...,...,...,...,...
10243,TECH-03,tech_park,2026-06-30 19:00:00,95,93,2,85,4.20,1.00
10244,TECH-03,tech_park,2026-06-30 20:00:00,92,90,2,62,2.72,1.05
10245,TECH-03,tech_park,2026-06-30 21:00:00,54,53,1,57,4.16,1.01
10246,TECH-03,tech_park,2026-06-30 22:00:00,34,33,1,32,3.81,1.02


In [14]:
airport_trips = pd.read_csv("../data/airport_trips.csv")
airport_trips

,trip_id,pickup_zone_id,drop_zone_id,drop_zone_type,request_ts,trip_distance_km,captain_cancelled,got_return_fare_within_20min,fare_inr
0,TRP00036169,APT-T1,CBD-02,city_core,2026-05-01 00:00:00,8.18,0,1,148.0
1,TRP00018764,APT-T1,SUB-07,suburban,2026-05-01 00:00:00,21.10,0,0,302.0
2,TRP00027350,APT-T2,TECH-03,tech_park,2026-05-01 00:00:00,14.22,0,0,231.0
3,TRP00058214,APT-T2,TECH-03,tech_park,2026-05-01 00:00:00,9.81,0,0,207.0
4,TRP00044709,APT-T1,SUB-07,suburban,2026-05-01 00:00:00,18.02,1,0,333.0
...,...,...,...,...,...,...,...,...,...
59995,TRP00038183,APT-T2,CBD-01,city_core,2026-06-30 23:00:00,22.22,0,1,346.0
59996,TRP00055072,APT-T1,CBD-02,city_core,2026-06-30 23:00:00,7.73,0,0,148.0
59997,TRP00018547,APT-T1,CBD-01,city_core,2026-06-30 23:00:00,16.58,0,1,264.0
59998,TRP00050675,APT-T1,SUB-07,suburban,2026-06-30 23:00:00,22.28,0,0,355.0


In [15]:
datasets = {
    "captains": captains,
    "doc_events": doc_events,
    "approvals": approvals,
    "activation": activation,
    "nudges": nudges,
    "airport_hourly": airport_hourly,
    "airport_trips": airport_trips
}

In [16]:
for name, df in datasets.items():
    print(f"{name}: {df.shape}")

captains: (25000, 9)
doc_events: (186282, 7)
approvals: (25000, 5)
activation: (4206, 5)
nudges: (16314, 6)
airport_hourly: (10248, 9)
airport_trips: (60000, 9)


## Dataset Inventory

In [17]:
inventory = pd.DataFrame({
    "dataset": datasets.keys(),
    "rows": [df.shape[0] for df in datasets.values()],
    "columns": [df.shape[1] for df in datasets.values()]
})

In [18]:
inventory

,dataset,rows,columns
0,captains,25000,9
1,doc_events,186282,7
2,approvals,25000,5
3,activation,4206,5
4,nudges,16314,6
5,airport_hourly,10248,9
6,airport_trips,60000,9


In [19]:
for name, df in datasets.items():
    print(f"\n{name}")
    print(df.columns.tolist())


captains
['captain_id', 'signup_ts', 'city', 'vehicle_type', 'acquisition_channel', 'signup_zone_id', 'device_tier', 'app_language', 'age_band']

doc_events
['event_id', 'captain_id', 'doc_type', 'attempt_no', 'event_type', 'event_ts', 'failure_reason']

approvals
['captain_id', 'decision_ts', 'final_status', 'last_stage_reached', 'docs_cleared']

activation
['captain_id', 'first_order_ts', 'orders_d7', 'orders_d30', 'online_hours_d30']

nudges
['captain_id', 'campaign_id', 'channel', 'sent_ts', 'delivered', 'clicked']

airport_hourly
['zone_id', 'zone_type', 'hour_ts', 'requests', 'fulfilled_requests', 'unfulfilled_requests', 'online_captains', 'avg_eta_min', 'avg_surge_multiplier']

airport_trips
['trip_id', 'pickup_zone_id', 'drop_zone_id', 'drop_zone_type', 'request_ts', 'trip_distance_km', 'captain_cancelled', 'got_return_fare_within_20min', 'fare_inr']


## Schema Inspection

In [21]:
for name, df in datasets.items():
    print(f"\n{'=' * 60}")
    print(name.upper())
    print("=" * 60)
    df.info()


CAPTAINS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   captain_id           25000 non-null  object
 1   signup_ts            25000 non-null  object
 2   city                 25000 non-null  object
 3   vehicle_type         25000 non-null  object
 4   acquisition_channel  25000 non-null  object
 5   signup_zone_id       25000 non-null  object
 6   device_tier          25000 non-null  object
 7   app_language         25000 non-null  object
 8   age_band             25000 non-null  object
dtypes: object(9)
memory usage: 1.7+ MB

DOC_EVENTS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186282 entries, 0 to 186281
Data columns (total 7 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   event_id        186282 non-null  object
 1   captain_id      186282 non-null  object
 

In [22]:
for name, df in datasets.items():
    print(f"\n{name}")
    display(df.head(3))


captains


,captain_id,signup_ts,city,vehicle_type,acquisition_channel,signup_zone_id,device_tier,app_language,age_band
0,CPT104359,2026-01-01 06:08:00,Pune,Auto,organic_app,PUN-Z23,low,kn,25-34
1,CPT120008,2026-01-01 06:52:00,Pune,Cab,referral,PUN-Z05,mid,mr,45+
2,CPT102049,2026-01-01 07:02:00,Hyderabad,Cab,organic_app,HYD-Z01,high,te,35-44



doc_events


,event_id,captain_id,doc_type,attempt_no,event_type,event_ts,failure_reason
0,EV000000000,CPT104359,DL,1,upload_success,2026-01-01 20:30:06.229672479,NaN
1,EV000000001,CPT104359,DL,1,verification_pass,2026-01-01 23:26:54.472703731,NaN
2,EV000000002,CPT104359,RC,1,upload_success,2026-01-02 09:36:49.336302871,NaN



approvals


,captain_id,decision_ts,final_status,last_stage_reached,docs_cleared
0,CPT104359,2026-01-06 08:11:48.775157583,approved,NaN,6
1,CPT120008,NaN,dropped_in_docs,AADHAAR,2
2,CPT102049,NaN,dropped_in_docs,PERMIT,3



activation


,captain_id,first_order_ts,orders_d7,orders_d30,online_hours_d30
0,CPT104359,2026-01-11 12:11:54.524199600,7.0,33.0,26.9
1,CPT105916,2026-01-08 04:16:11.743250188,15.0,48.0,29.1
2,CPT115782,2026-01-10 11:14:57.939110321,9.0,31.0,20.2



nudges


,captain_id,campaign_id,channel,sent_ts,delivered,clicked
0,CPT104359,CAMP_SMS_004,sms,2026-01-02 01:31:34.373244464,1,0
1,CPT102049,CAMP_WA_002,whatsapp,2026-01-02 20:02:31.634258914,1,0
2,CPT123006,CAMP_WA_002,whatsapp,2026-01-03 11:32:05.154545929,1,1



airport_hourly


,zone_id,zone_type,hour_ts,requests,fulfilled_requests,unfulfilled_requests,online_captains,avg_eta_min,avg_surge_multiplier
0,APT-T1,airport_terminal,2026-05-01 00:00:00,83,31,52,15,9.13,2.04
1,APT-T1,airport_terminal,2026-05-01 01:00:00,111,27,84,10,10.13,2.17
2,APT-T1,airport_terminal,2026-05-01 02:00:00,86,21,65,13,10.89,2.19



airport_trips


,trip_id,pickup_zone_id,drop_zone_id,drop_zone_type,request_ts,trip_distance_km,captain_cancelled,got_return_fare_within_20min,fare_inr
0,TRP00036169,APT-T1,CBD-02,city_core,2026-05-01 00:00:00,8.18,0,1,148.0
1,TRP00018764,APT-T1,SUB-07,suburban,2026-05-01 00:00:00,21.10,0,0,302.0
2,TRP00027350,APT-T2,TECH-03,tech_park,2026-05-01 00:00:00,14.22,0,0,231.0


## Grain and Key Checks

In [23]:
key_checks = {
    "captains": ["captain_id"],
    "doc_events": ["event_id"],
    "approvals": ["captain_id"],
    "activation": ["captain_id"],
    "airport_trips": ["trip_id"]
}

In [24]:
for name, keys in key_checks.items():
    df = datasets[name]
    
    print(f"\n{name}")
    print(f"Rows: {len(df):,}")
    print(f"Unique keys: {df[keys].drop_duplicates().shape[0]:,}")


captains
Rows: 25,000
Unique keys: 25,000

doc_events
Rows: 186,282
Unique keys: 186,282

approvals
Rows: 25,000
Unique keys: 25,000

activation
Rows: 4,206
Unique keys: 4,206

airport_trips
Rows: 60,000
Unique keys: 60,000


In [25]:
print(nudges.groupby(["captain_id", "campaign_id"]).size().describe())

count    16314.0
mean         1.0
std          0.0
min          1.0
25%          1.0
50%          1.0
75%          1.0
max          1.0
dtype: float64


In [26]:
print(
    airport_hourly.groupby(["zone_id", "hour_ts"])
    .size()
    .value_counts()
    .sort_index()
)

1    10248
Name: count, dtype: int64


## Categorical Variables

In [27]:
categorical_columns = {
    "captains": [
        "city",
        "vehicle_type",
        "acquisition_channel",
        "device_tier",
        "app_language",
        "age_band"
    ],
    "doc_events": [
        "doc_type",
        "event_type",
        "failure_reason"
    ],
    "approvals": [
        "final_status",
        "last_stage_reached"
    ],
    "nudges": [
        "campaign_id",
        "channel"
    ],
    "airport_hourly": [
        "zone_type"
    ],
    "airport_trips": [
        "drop_zone_type"
    ]
}

In [28]:
for dataset, columns in categorical_columns.items():
    print(f"\n{'=' * 60}")
    print(dataset.upper())
    print("=" * 60)
    
    df = datasets[dataset]
    
    for column in columns:
        print(f"\n{column}:")
        print(df[column].value_counts(dropna=False))


CAPTAINS

city:
city
Hyderabad    7663
Delhi        6452
Pune         5448
Bangalore    5437
Name: count, dtype: int64

vehicle_type:
vehicle_type
Auto         11363
Cab           8881
ERickshaw     4756
Name: count, dtype: int64

acquisition_channel:
acquisition_channel
organic_app       8418
fos_field         6122
referral          5518
paid_digital      3483
gc_telecalling    1459
Name: count, dtype: int64

device_tier:
device_tier
low     11565
mid      9533
high     3902
Name: count, dtype: int64

app_language:
app_language
kn    5081
te    5064
mr    4956
en    4952
hi    4947
Name: count, dtype: int64

age_band:
age_band
25-34    10397
35-44     7086
18-24     4540
45+       2977
Name: count, dtype: int64

DOC_EVENTS

doc_type:
doc_type
DL           50780
RC           48099
AADHAAR      31117
FITNESS      21884
PERMIT       20807
INSURANCE    13595
Name: count, dtype: int64

event_type:
event_type
upload_success       93236
verification_pass    73274
verification_fail    19772


## Timestamp Inventory

In [30]:
timestamp_columns = {
    "captains": ["signup_ts"],
    "doc_events": ["event_ts"],
    "approvals": ["decision_ts"],
    "activation": ["first_order_ts"],
    "nudges": ["sent_ts"],
    "airport_hourly": ["hour_ts"],
    "airport_trips": ["request_ts"]
}

In [31]:
for dataset, columns in timestamp_columns.items():
    print(f"\n{dataset}")
    
    for column in columns:
        values = pd.to_datetime(datasets[dataset][column], errors="coerce")
        
        print(
            f"{column}: "
            f"min={values.min()} | "
            f"max={values.max()} | "
            f"null_after_parse={values.isna().sum():,}"
        )


captains
signup_ts: min=2026-01-01 06:08:00 | max=2026-06-30 21:57:00 | null_after_parse=0

doc_events
event_ts: min=2026-01-01 13:30:50.795752785 | max=2026-06-30 23:56:42.723769018 | null_after_parse=0

approvals
decision_ts: min=2026-01-05 21:40:24.347261353 | max=2026-06-30 22:26:04.692838376 | null_after_parse=20,359

activation
first_order_ts: min=2026-01-06 11:57:58.804450784 | max=2026-06-30 21:36:06.089121524 | null_after_parse=102

nudges
sent_ts: min=2026-01-01 23:10:49.644254087 | max=2026-06-30 23:37:42.004031897 | null_after_parse=0

airport_hourly
hour_ts: min=2026-05-01 00:00:00 | max=2026-06-30 23:00:00 | null_after_parse=0

airport_trips
request_ts: min=2026-05-01 00:00:00 | max=2026-06-30 23:00:00 | null_after_parse=0


## Numerical Sanity Checks

In [32]:
for name, df in datasets.items():
    print(f"\n{'=' * 60}")
    print(name.upper())
    print("=" * 60)
    
    numeric_cols = df.select_dtypes(include="number").columns
    
    if len(numeric_cols) > 0:
        display(df[numeric_cols].describe().T)


CAPTAINS

DOC_EVENTS


,count,mean,std,min,25%,50%,75%,max
attempt_no,186282.0,1.141447,0.381474,1.0,1.0,1.0,1.0,3.0



APPROVALS


,count,mean,std,min,25%,50%,75%,max
docs_cleared,25000.0,2.9814,2.034822,0.0,1.0,3.0,5.0,6.0



ACTIVATION


,count,mean,std,min,25%,50%,75%,max
orders_d7,4019.0,7.068674,4.508138,0.0,4.0,6.0,9.000,37.0
orders_d30,3372.0,27.531435,17.011344,0.0,16.0,24.0,35.000,161.0
online_hours_d30,3372.0,20.731346,13.395880,0.0,11.7,18.2,26.225,117.8



NUDGES


,count,mean,std,min,25%,50%,75%,max
delivered,16314.0,0.916207,0.277086,0.0,1.0,1.0,1.0,1.0
clicked,16314.0,0.335969,0.472343,0.0,0.0,0.0,1.0,1.0



AIRPORT_HOURLY


,count,mean,std,min,25%,50%,75%,max
requests,10248.0,44.653591,21.222099,3.0,29.0,40.00,55.00,132.00
fulfilled_requests,10248.0,38.375195,17.750041,2.0,25.0,35.00,47.00,119.00
unfulfilled_requests,10248.0,6.278396,15.866283,1.0,1.0,1.00,1.00,99.00
online_captains,10248.0,36.251756,16.593118,2.0,25.0,34.00,46.00,106.00
avg_eta_min,10248.0,4.267979,1.869872,2.0,3.2,3.80,4.53,12.60
avg_surge_multiplier,10248.0,1.145376,0.302356,1.0,1.0,1.04,1.09,2.36



AIRPORT_TRIPS


,count,mean,std,min,25%,50%,75%,max
trip_distance_km,60000.0,17.883161,8.117206,1.54,11.82,16.62,22.66,66.73
captain_cancelled,60000.0,0.137667,0.344553,0.00,0.00,0.00,0.00,1.00
got_return_fare_within_20min,60000.0,0.359600,0.479887,0.00,0.00,0.00,1.00,1.00
fare_inr,60000.0,279.349983,111.685797,50.00,197.00,263.00,345.00,931.00


In [33]:
airport_hourly["check_fulfillment"] = (
    airport_hourly["fulfilled_requests"]
    + airport_hourly["unfulfilled_requests"]
    - airport_hourly["requests"]
)

airport_hourly["check_fulfillment"].value_counts(dropna=False)

check_fulfillment
0    10248
Name: count, dtype: int64

In [34]:
airport_hourly.drop(columns="check_fulfillment", inplace=True)

In [35]:
activation[
    activation["orders_d7"] > activation["orders_d30"]
][
    ["captain_id", "orders_d7", "orders_d30"]
]

,captain_id,orders_d7,orders_d30
1525,CPT123585,4.0,3.0
1919,CPT123478,4.0,3.0
2048,CPT103607,7.0,6.0
2665,CPT124543,10.0,7.0
2770,CPT112526,6.0,5.0
3087,CPT116344,6.0,5.0
3347,CPT105360,4.0,3.0


In [36]:
airport_trips[
    ["trip_distance_km", "fare_inr"]
].describe().T

,count,mean,std,min,25%,50%,75%,max
trip_distance_km,60000.0,17.883161,8.117206,1.54,11.82,16.62,22.66,66.73
fare_inr,60000.0,279.349983,111.685797,50.00,197.00,263.00,345.00,931.00


In [37]:
airport_trips[
    ["trip_distance_km", "fare_inr"]
].isna().sum()

trip_distance_km    0
fare_inr            0
dtype: int64

In [38]:
for name, df in datasets.items():
    numeric_cols = df.select_dtypes(include="number").columns
    
    if len(numeric_cols) > 0:
        negatives = (df[numeric_cols] < 0).sum()
        negatives = negatives[negatives > 0]
        
        if len(negatives):
            print(name)
            print(negatives)

## Missing Value Overview

In [39]:
missing_summary = []

for name, df in datasets.items():
    for column in df.columns:
        missing_count = df[column].isna().sum()
        
        if missing_count > 0:
            missing_summary.append({
                "dataset": name,
                "column": column,
                "missing_count": missing_count,
                "missing_pct": round(100 * missing_count / len(df), 2)
            })

missing_summary = pd.DataFrame(missing_summary)

missing_summary.sort_values(
    ["dataset", "missing_pct"],
    ascending=[True, False]
)

,dataset,column,missing_count,missing_pct
5,activation,orders_d30,834,19.83
6,activation,online_hours_d30,834,19.83
4,activation,orders_d7,187,4.45
3,activation,first_order_ts,102,2.43
1,approvals,decision_ts,20359,81.44
2,approvals,last_stage_reached,4393,17.57
0,doc_events,failure_reason,166510,89.39


## Referential Integrity Checks

In [40]:
captain_ids = set(captains["captain_id"])

checks = {
    "doc_events captain_id": set(doc_events["captain_id"]).issubset(captain_ids),
    "approvals captain_id": set(approvals["captain_id"]).issubset(captain_ids),
    "activation captain_id": set(activation["captain_id"]).issubset(captain_ids),
    "nudges captain_id": set(nudges["captain_id"]).issubset(captain_ids),
}

for check, result in checks.items():
    print(f"{check}: {result}")

approved_ids = set(
    approvals.loc[
        approvals["final_status"] == "approved",
        "captain_id"
    ]
)

activation_ids = set(activation["captain_id"])

print(
    "Activation subset of approved captains:",
    activation_ids.issubset(approved_ids)
)

doc_events captain_id: True
approvals captain_id: True
activation captain_id: True
nudges captain_id: True
Activation subset of approved captains: True


## Extraction Cutoff Check

In [42]:
cutoff = pd.Timestamp("2026-06-30 23:59:59")

timestamp_columns_for_cutoff = {
    "captains": "signup_ts",
    "doc_events": "event_ts",
    "approvals": "decision_ts",
    "activation": "first_order_ts",
    "nudges": "sent_ts",
    "airport_hourly": "hour_ts",
    "airport_trips": "request_ts"
}

for dataset, column in timestamp_columns_for_cutoff.items():
    ts = pd.to_datetime(datasets[dataset][column], errors="coerce")
    
    print(
        f"{dataset}:",
        (ts > cutoff).sum(),
        "records after cutoff"
    )

captains: 0 records after cutoff
doc_events: 0 records after cutoff
approvals: 0 records after cutoff
activation: 0 records after cutoff
nudges: 0 records after cutoff
airport_hourly: 0 records after cutoff
airport_trips: 0 records after cutoff


## Failure Reason Consistency

In [43]:
failure_reason_check = doc_events[
    (
        doc_events["failure_reason"].notna()
        & (doc_events["event_type"] != "verification_fail")
    )
    |
    (
        doc_events["failure_reason"].isna()
        & (doc_events["event_type"] == "verification_fail")
    )
]

print(
    "Rows with inconsistent failure_reason/event_type:",
    len(failure_reason_check)
)

Rows with inconsistent failure_reason/event_type: 0


## Nudge Delivery / Click Consistency

In [44]:
invalid_nudges = nudges[
    (nudges["clicked"] == 1)
    & (nudges["delivered"] == 0)
]

print(
    "Clicks without delivery:",
    len(invalid_nudges)
)

Clicks without delivery: 457


## Document Attempt Validation

In [45]:
invalid_attempts = doc_events[
    ~doc_events["attempt_no"].between(1, 3)
]

print(
    "Events with attempt_no outside 1-3:",
    len(invalid_attempts)
)

Events with attempt_no outside 1-3: 0


In [46]:
attempt_distribution = (
    doc_events["attempt_no"]
    .value_counts()
    .sort_index()
)

attempt_distribution

attempt_no
1    162176
2     21863
3      2243
Name: count, dtype: int64

## Initial Observations

### Dataset Structure

I have seven datasets in total. Five of them are related to captain onboarding and activation, while the other two are related to airport supply.

The onboarding datasets are `captains`, `doc_events`, `approvals`, `activation`, and `nudges`. The airport datasets are `airport_hourly` and `airport_trips`.

The sizes are quite different across the datasets. `captains` and `approvals` both have 25,000 rows, while `doc_events` has 186,282 rows because one captain can generate multiple document events. `activation` has 4,206 rows and represents approved captains. `nudges` has 16,314 records. The airport datasets contain 10,248 hourly zone records and 60,000 sampled trips.

The datasets are not all at the same level of detail, so I should be careful with joins later. In particular, document events and nudges cannot be joined directly to the captain table without first deciding what captain-level information I need from them.

### Dataset Grain

`captains` has one row for each signup and `captain_id` is unique.

`doc_events` has one row for each document event and `event_id` is unique. A captain can appear many times here because a document can be uploaded and verified more than once.

`approvals` has one row per captain and contains the final onboarding status.

`activation` contains approved captains and has one row per captain. The activation records are all linked to captains who have an approved status.

`nudges` contains communication records for captains and campaigns. A captain can therefore have more than one record.

`airport_hourly` has one record for each zone and hour. The zone and hour combination does not repeat.

`airport_trips` has one record per sampled airport trip and `trip_id` is unique.

Because of these different grains, I will keep the onboarding analysis at captain level where possible and aggregate event-level data before joining it to captain-level data.

### Timestamp Observations

The captain signup data starts in January 2026 and runs until June 30, 2026. The document events, approvals, activation and nudge data also fall within this period.

The airport data only covers May and June 2026. This is important because the airport analysis has a shorter observation period than the onboarding analysis.

The timestamps could be parsed successfully and I did not find invalid dates.

I also checked the extraction cutoff mentioned in the assignment. There are no records after June 30, 2026 at 23:59:59 in any of the seven datasets.

For the analysis, I will use the cutoff from the assignment consistently and will be careful when comparing recent signup cohorts because newer captains may not have had enough time to complete onboarding or become active.

### Categorical Observations

The captain data covers four cities: Bangalore, Delhi, Hyderabad and Pune.

There are three vehicle types: Auto, Cab and ERickshaw. The acquisition channels include organic app, field sales, referrals, paid digital and telecalling.

The data also contains different device tiers, app languages and age bands, which can be useful later when looking for differences in onboarding completion.

The document events contain six document types and three event types: upload success, verification pass and verification fail.

The approval data has four outcomes: approved, rejected, dropped in docs and in progress.

For the airport data, there are four zone types including airport terminals, city areas, suburban areas and tech parks.

The document order described in the assignment will matter when analysing the onboarding funnel. Permit also needs to be treated differently because it only applies to Auto and Cab.

### Potential Data Quality Issues

There are some missing values, but not all of them appear to be actual data problems.

`failure_reason` is missing for most document events. This makes sense because a failure reason is mainly relevant when a verification fails. I checked the relationship between `failure_reason` and `event_type` and did not find any inconsistent rows.

`decision_ts` is missing for a large part of the approvals data. This appears to be related to captains who are still in progress or dropped during onboarding, so I will not fill these values without understanding the status first.

There are 102 approved captains without a `first_order_ts`. This is useful for the later activation analysis because approval does not necessarily mean that the captain actually becomes active.

Some activation records have missing D7 and D30 metrics. These may be caused by recent records not having a complete observation window, so I need to account for this when creating cohorts.

There are seven records where `orders_d7` is greater than `orders_d30`. Since D30 includes a longer period than D7, these records look unusual and should be investigated before using these fields.

There are also 457 nudge records where a click is recorded even though delivery is marked as false. I will keep these records for now and treat them as a data quality issue rather than changing them without knowing why they occur.

The document attempt values are all between 1 and 3, so there are no attempts outside the limit mentioned in the assignment.

The airport request numbers also pass the basic consistency check because fulfilled requests plus unfulfilled requests equals total requests.

The captain IDs in the event, approval, activation and nudge datasets all match records in the captain table. The activation records also belong only to approved captains.

### Questions to Investigate in Data Audit

The main things I want to check next are whether the document events follow the expected onboarding order and whether the attempts and verification results make sense for each captain.

I also want to understand where captains are dropping out of the onboarding process and whether the drop off is concentrated in a particular document or failure reason.

After that, I will compare onboarding performance across useful groups such as city, vehicle type, acquisition channel and device tier. I need to be careful not to treat a difference between groups as proof that one factor caused the difference.

Another important check is cohort maturity. Recent signups may still be going through onboarding, so using all signups together could make the funnel look worse than it actually is.

For activation, I want to understand why some approved captains do not have a first order and whether the missing D7 and D30 values are explained by the age of the cohort.

I also need to investigate the unusual activation records and the nudge records where clicks are present without delivery.

For the campaign analysis, I will need to check whether captains who received `CAMP_WA_002` were actually comparable to those who did not receive it. A higher completion rate by itself will not prove that the campaign caused the improvement.

For the airport analysis, I want to identify when the demand and supply gap is largest, how persistent it is, and what happens after an airport trip. This will help determine whether the problem is really a shortage of captains or whether there is another operational reason behind the unmet demand.
